x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java

```diff
@@ -1643,6 +1671,137 @@ else if ((other instanceof LessThan || other instanceof LessThanOrEqual) &&
+                }
+            }
+
+            // check on "condition-overlapping" inequalities:
+            // a != 2 AND a > 3 -> a > 3 (discard NotEquals)
+            // a != 2 AND a >= 2 -> a > 2 (discard NotEquals plus update inequality)
+            // a != 2 AND a > 1 -> nop (do nothing)
+            //
+            // a != 2 AND a < 3 -> nop
+            // a != 2 AND a <= 2 -> a < 2
+            // a != 2 AND a < 1 -> a < 1
+            for (int i = 0; i < bcs.size(); i ++) {
+                BinaryComparison bc = bcs.get(i);
+
+                if (bc instanceof LessThan || bc instanceof LessThanOrEqual) {
+                    comp = bc.right().foldable() ? BinaryComparison.compare(neqVal, bc.right().fold()) : null;
+                    if (comp != null) {
+                        if (comp >= 0) {
+                            if (comp == 0 && bc instanceof LessThanOrEqual) { // a != 2 AND a <= 2 -> a < 2
+                                bcs.set(i, new LessThan(bc.source(), bc.left(), bc.right()));
+                            } // else : comp > 0 (a != 2 AND a </<= 1 -> a </<= 1), or == 0 && bc i.of "<" (a != 2 AND a < 2 -> a < 2)
+                            return true;
+                        } // else: comp < 0 : a != 2 AND a </<= 3 -> nop
+                    } // else: non-comparable, nop
+                } else if (bc instanceof GreaterThan || bc instanceof GreaterThanOrEqual) {
+                    comp = bc.right().foldable() ? BinaryComparison.compare(neqVal, bc.right().fold()) : null;
+                    if (comp != null) {
+                        if (comp <= 0) {
+                            if (comp == 0 && bc instanceof GreaterThanOrEqual) { // a != 2 AND a >= 2 -> a > 2
+                                bcs.set(i, new GreaterThan(bc.source(), bc.left(), bc.right()));
+                            } // else: comp < 0 (a != 2 AND a >/>= 3 -> a >/>= 3), or == 0 && bc i.of ">" (a != 2 AND a > 2 -> a > 2)
+                            return true;
+                        } // else: comp > 0 : a != 2 AND a >/>= 1 -> nop
+                    } // else: non-comparable, nop
+                } // else: other non-relevant type
+            }
+
+            return false;
+        }
+
+        // a != X OR Y < a < Z -> a != X (plus equivalent inequalities)
+        // a != X OR X <= a < Z -> TRUE (plus equality on upper bound, plus equivalent inequalities)
+        private static Boolean filterDisjunctionByNotEquals(List<NotEquals> notEquals, List<BinaryComparison> bcs, List<Range> ranges) {
+            Boolean updated = false;
+
+            for (NotEquals neq : notEquals) {
+                for (Iterator<BinaryComparison> bcIterator = bcs.iterator(); bcIterator.hasNext(); ) {
+                    BinaryComparison bc = bcIterator.next();
+                    if (neq.left().semanticEquals(bc.left())) {
+                        if (bc instanceof LessThan || bc instanceof GreaterThan) { // a != 2 OR a < 3 -> a != 2 (plus LessThen)
```

Comment:
I don't understand this simplification here. If the not-equality and a `<` or `>` act on the same variable (`a` in this case`) why is `a != 2 OR a < 3` resulting in a `a != 2` condition?

One edge case is `a = 2` and in this case it matches `a < 3`. The other edge case is `anything >= 3` and in this case it matches the condition `a != 2`. The way I see it, `a != 2 OR a < 3` is always TRUE.

BUT, in Elasticsearch these two conditions being existent in a query (even if they can be simplified), they can contribute to the scoring of a document: an `OR` is translated into a [`bool` query with two `should` statements](https://www.elastic.co/guide/en/elasticsearch/reference/7.6/query-dsl-bool-query.html). And, even if, the documents being returned are the same in a simplified (assuming the simplification returns always TRUE) and non-simplified way, if the user also asks for SCORE(), the actual results will be different in the two cases. I would take a step back and re-evaluate the `OR` simplification in general to make sure it doesn't affect the scoring of documents, in case SCORE() is returned.

Reply:
> One edge case is `a = 2` and in this case it matches `a < 3`. The other edge case is `anything >= 3` and in this case it matches the condition `a != 2`. The way I see it, `a != 2 OR a < 3` is always `TRUE`.

That's actually `TRUE`, thanks for catching it!
I believe the correct optimisation for `a != X OR Y </<= a </<= Z` should be to:
* `a != X` if:
   * X ∉ (Y, Z); or
   * bound equalities are disjunctive: `X==Y && (Y < a ... )` or `X==Z && (... a < Z)`; 
* `TRUE` in all other cases.

And same goes for inequalities, where Y or Z would become -∞ or +∞ respectively.
I'll fix it.

Reply:
> I would take a step back and re-evaluate the OR simplification in general to make sure it doesn't affect the scoring of documents, in case SCORE() is returned.

Interesting point. Do you think it'd be easy (or even necessary) to come up with an example? This being about numerical conditions optimisations, I'm wondering if the scoring would be indeed affected, in isolation: `a != 2 OR a > 3 OR b = 'foo'` vs. `a != 2 OR b = 'foo'` would yield different scores?
I'll look into it (unless you're positive about it and I can save that time). Others - @costin? - welcome to chime in, I believe the simplifications in the optimiser  are not isolated to equalities and inequalities around `AND`/`OR`s.

Reply:
@bpintea `select SCORE(), salary from test_emp where salary <> 70000 OR salary < 30000 ORDER BY salary`

Reply:
Thanks.
It seems indeed that other (existing) optimisations would also impact the scoring. (Smth like `WHERE salary < 30000 OR salary < 50000` yield same scores as without the first condition, merging inequalities into ranges would probably also affect it etc.)

So I guess the presence of scoring might require a deeper consideration in regards to deploying the optimiser?

Reply:
@astefan has a good point. The optimizations for `AND` should be safe since either you pass all o them or none - thus there's no scoring.
However with `OR` one result can match multiple queries hence why I would trim the optimizations to a minimum.
We could add awarness regarding scoring and thus enable some optimizations only when no scoring is needed however I think that adds additional complexity for unclear gains thus I would spend the effort somewhere else.

Reply:
> However with OR one result can match multiple queries hence why I would trim the optimizations to a minimum.

Sure, I'll then trim the extras.

> We could add awarness regarding scoring and thus enable some optimizations only when no scoring is needed however I think that adds additional complexity for unclear gains thus I would spend the effort somewhere else.

Got it. I guess machine-generated queries - maybe also BIs - can produce very verbose statements, but it's true that the optimisations can be an obscure ground and maybe not worth it if users tend to go for the scores as well.

Would opening a ticket to track/remove existing optimisation be worth it?
https://github.com/elastic/elasticsearch/blob/9396e7574f608bb8262a9f4863c4dc637271baac/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java#L1367
https://github.com/elastic/elasticsearch/blob/9396e7574f608bb8262a9f4863c4dc637271baac/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java#L1435
potentially others?

Reply:
> Would opening a ticket to track/remove existing optimisation be worth it?

Please go ahead - it's better to have an issue that we close than forget about it because there's no issue.